src/groupeddataframe/utils.jl

```diff
@@ -321,9 +325,15 @@ function row_group_slots(cols::NTuple{N, AbstractVector},
 
     lg = length(groups)
     nt = Threads.nthreads()
-    # disable threading if we are processing a small data frame or number of groups is large
-    if lg < 1_000_000 || ngroups > lg * (0.5 - 1 / (2 * nt)) / (2 * nt)
-       nt = 1
+    # make sure we are processing at least 100_000 rows per task if we do threading
+    if lg < 100_000 * nt
+       nt = max(1, lg ÷ 100_000)
+    end
+    # if there are many groups limit the number of threads used
+    if 16 * Int(ngroups) > lg
+        nt = 1
+    else
+        nt = clamp(floor(Int, lg * (sqrt(max(0.0, 1 - 16 * (ngroups / lg))) + 1) / (8 * ngroups)), 2, nt)
```

Comment:
How did you choose this formula? I don't remember how we chose the previous one either.

Reply:
This is the same formula as we previously had `ngroups > lg * (0.5 - 1 / (2 * nt)) / (2 * nt)`, just solved for `nt`.
Some time ago we manually tried to find the best formula.
The tension here is:
* first twe process chunks in `nt` threads; each chunk produces a vector (so we prefer to have many chunks)
* then we need to aggregate these `nt` vectors into one vector (so we do not want to have too many chunks)

So there is a tension. However, I have analyzed the formulas and they can be simplified to just:
```
lg / ngroups / 4 - 2
```
(and `lg/ngroups` is group size - if groups are larger we allow for more threads)

I made this change.

Reply:
OK. Wouldn't it make sense to use a single formula for all cases, like this (without the `if`)?
```
    nt = clamp(floor(Int, (lg / 4) / ngroups - 2), 1, nt)
```

Reply:
You are right. This is equivalent. Fixed.